### PR TITLE
Make float16 actually work using gradient and optimizer state scaling

### PR DESCRIPTION
**Pull request by e-c-d**
Float16 training doesn't actually work because the gradients underflow. Simply applying `torch.amp.GradScaler` doesn't fix the problem because the optimizer state variables are then either too small (causing underflow) or too large (causing overflow). Sometimes there is no single gradient scaling factor that avoids both underflow and overflow, owing to the limited dynamic range of float16. For example, Adam's state has exponential moving averages of both `grad` and `grad^2`. If you scale up `grad` to a healthy value around 1000, then the `grad^2` will be 1,000,000 which is far beyond the maximum representable float16 which is 65504.

This PR enables both gradient scaling and automatic optimizer state scaling for Adam and AdamW, except when using foreach or when using coupled weight decay. Only minimal changes are required to optimizers. I will extend this soon to a few other optimizers (StableAdamW and Adan). Another important change is that moving the weight decay to be inside the Kahan summation in order to avoid rounding error.

I figured I would clean up a bit what already works for me and contribute it back to the community.

### CODE REVIEW

**Comment by e-c-d** [comment]
I also implemented gradient scaling for the StableAdamW and Adan triton backends. Float16 test cases now actually pass for them. I didn't even have to decrease the number of iterations for Adan, but I did have to increase the error tolerances slightly. Anyway, I think I'll take a break for now.

Thank you for this wonderful library and test suite, it's been delightful to read and work with. My apologies for the mess I've made of the code.

**Comment by warner-benjamin** [comment]
@e-c-d  Thanks for the PR. Busy atm but will take a look at it next week.
